docs/src/references/agent/operation-workflow.md

```diff
@@ -0,0 +1,259 @@
+  script = "/usr/bin/firmware_handler.sh rollback"
+  next = ["rollback_reboot"]
+
+[rollback_reboot]
+  next = ["failed"]
+
+[successful]
+  next = []
+
+[failed]
+  next = []
+```
+
+Thin-edge combines all these workflows to determine what has to be done
+when a state message is published for a command on a topic matching the global topic filter for commands,
+i.e. `te/+/+/+/+/cmd/+/+`.
+- Each running instance of the __tedge_agent__ reacts only on commands targeting its own device.
+- If a user-defined workflow has been defined for this operation, then this workflow is used to determine the required action.
+- If no workflow has been defined for the user for this operation, then the built-in workflow is used.
+- If there is no workflow or no defined action for the current state, then the __tedge_agent__ fails the command.
+
+### Built-in Actions
+
+For operations supported by the __tedge-agent__, the built-in behavior is the default.
+In other words, if no alternative actions is specified by a user-provided workflow,
+the commands are processed following the built-in behavior.
+
+If alternative actions are only given on some states, then the built-in behavior is applied to all others.
+This gives the ability to override parts of the built-in operation workflow.
+
+The __successful__ and __failed__ states are also handled in a specific way.
+Only the command issuer is supposed to react on those,
+pushing a retained empty message on the command request topic when done.
+This ends the command workflow.
+
+### Script Execution
+
+A script can be attached to a command state. 
+
+```
+[state]
+script = "/full/path/command [args]" 
+```
+
+This script is given as a plain command line possibly with arguments.
+
+Data extracted from the command status topic and message payload can be passed as argument to the script.
+
+- `"/bin/new-command.sh ${.topic} ${.payload}"` passes two arguments to the `/bin/new-command.sh` program.
+  - The first one is the full command request topic (*e.g.* `te/device/main/cmd///restart/c8y-mapper-123`).
+  - The second one is the full json payload (*e.g.* `{"status': "init"}`).
+- Specific path expressions can be used to pass specific excerpts.
+  - `${.}` is a json for the whole message including the `topic` and the `payload`.
+  - `${.topic}` is the command request topic (*e.g.* `te/device/main/cmd///restart/c8y-mapper-123`)
+  - `${.topic.target}` is the command target identity  (*e.g.* `device/main/cmd//`)
+  - `${.topic.operation}` is the command operation  (*e.g.* `restart`)
+  - `${.topic.cmd_id}` is the command request unique identifier  (*e.g.* `c8y-mapper-123`)
+  - `${.payload}` is the whole command json payload (*e.g.* `{"status': "init"}`)
+  - `${.payload.status}` is the command current status (*e.g.* `"init"`)
+  - `${.payload.x.y.z}` is the json value extracted from the payload following the given `x.y.z` path if any.
+  - If given `${.some.unknown.path}`, the argument is passed unchanged to the script.
```

Comment:
Shouldn't that be a failure, as it indicates an unexpected JSON output, which doesn't conform to the format that the script expected?

Reply:
By passing the value to the script at least it gives the script a chance to run some custom logic (e.g. cleaning up, print a more usable error message etc.).

We could look at adding a "strict" mode later if needed.